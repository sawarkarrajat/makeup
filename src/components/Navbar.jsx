import React from "react";
/**
 * @property {Function} - generates a pure functional component for header
 */
export default function Navbar() {
  return (
    <div className="dashboard__navcontainer">
      <div className="dashboard__title">
        <h1>Makeup</h1>
      </div>
      <div className="dashboard__menu">
        <span id="mtxt">brands, lipsticks, nailpolish and much more!!!</span>
      </div>
    </div>
  );
}
