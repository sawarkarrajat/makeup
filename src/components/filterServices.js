export const ratingFilter = (products, rating) => {
  let data = products.filter((item) => {
    return item.rating === rating || item.rating === null;
  });
  console.log("value in rating filter", data);
  return data;
};
