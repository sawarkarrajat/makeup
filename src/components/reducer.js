export const initialState = {
  brandFiltersArray: [],
  tagFiltersArray: [],
  clearFilter: false,
  priceMin: null,
  priceMax: null,
  rating: null,
};

const reducer = (state, action) => {
  switch (action.type) {
    case "ADD_TO_BRAND_FILTER":
      return {
        ...state,
        brandFiltersArray: [...state.brandFiltersArray, action.item.label],
      };

    case "REMOVE_FROM_BRAND_FILTER":
      return {
        ...state,
        brandFiltersArray: [
          ...state.brandFiltersArray.filter(
            (element) => element !== action.item.label
          ),
        ],
      };
    case "ADD_TO_TAG_FILTER":
      return {
        ...state,
        tagFiltersArray: [...state.tagFiltersArray, action.item.label],
      };

    case "REMOVE_FROM_TAG_FILTER":
      return {
        ...state,
        tagFiltersArray: [
          ...state.tagFiltersArray.filter(
            (element) => element !== action.item.label
          ),
        ],
      };
    case "ADD_PRICE_RANGE":
      return {
        ...state,
        priceMin: action.item.min,
        priceMax: action.item.max,
      };
    case "UPDATE_RATING":
      return {
        ...state,
        rating: action.item,
      };
    case "CLEAR_FILTER":
      return {
        brandFiltersArray: [],
        tagFiltersArray: [],
        priceMin: null,
        priceMax: null,
        clearFilter: !state.clearFilter,
      };

    default:
      return state;
  }
};

export default reducer;
