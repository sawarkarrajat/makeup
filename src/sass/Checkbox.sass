@import base.sass
:root 
  --color: white
  --disabled: #959495

  
.checkbox 
  display: grid
  align-items: center
  grid-template-columns: min-content auto
  grid-gap: 10px
  font-size: 1em
  margin: 4px 0
  color: var(--color)

  &--disabled 
    color: var(--disabled)
  


.checkbox__control 
  display: inline-grid
  height: 20px
  align-items: center
  width: 1em
  height: 1em
  border-radius: 0.25em
  border: 0.1em solid white

  svg 
    transition: transform 0.1s ease-in 25ms
    transform: scale(0)
    transform-origin: bottom left
  


.checkbox__input 
  display: grid
  align-items: center
  grid-template-areas: "checkbox"

  > * 
    grid-area: checkbox
  

  input 
    opacity: 0
    width: 1em
    height: 1em

    &:hover + .checkbox__control 
      box-shadow: 0 0 0 0.05em #fff, 0 0 0.15em 0.1em $clr2
    

    &:checked + .checkbox__control svg 
      transform: scale(1)
    

    &:disabled + .checkbox__control 
      color: var(--disabled)
    
.radio__label
  // padding-top: 5px
  color: white

