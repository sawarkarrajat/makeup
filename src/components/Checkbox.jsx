import React, { useState, useEffect } from "react";
import "../sass/Checkbox.sass";
import { useStateValue } from "./StateProvider";

function Checkbox({ filterLabel, label }) {
  const [check, setCheck] = useState(false);
  const [
    { brandFiltersArray, tagFiltersArray, clearFilter },
    dispatch,
  ] = useStateValue();
  const handleChange = (event) => {
    let item = {
      label: label,
    };
    console.log("event.target.checked", event.target.checked);
    setCheck(!check);
    if (event.target.checked) {
      if (filterLabel === "brands") {
        dispatch({
          type: "ADD_TO_BRAND_FILTER",
          item: item,
        });
      } else {
        dispatch({
          type: "ADD_TO_TAG_FILTER",
          item: item,
        });
      }
    } else {
      if (filterLabel === "brands") {
        dispatch({
          type: "REMOVE_FROM_BRAND_FILTER",
          item: item,
        });
      } else {
        dispatch({
          type: "REMOVE_FROM_TAG_FILTER",
          item: item,
        });
      }
    }
    console.log("value in state", brandFiltersArray, tagFiltersArray);
  };
  useEffect(() => {
    if (brandFiltersArray.includes(label) || tagFiltersArray.includes(label)) {
      console.log(
        "contains brand/tag",
        brandFiltersArray.includes(label),
        tagFiltersArray.includes(label)
      );
      setCheck(true);
    } else {
      setCheck(false);
    }
  }, [brandFiltersArray, tagFiltersArray, label]);
  useEffect(() => {
    setCheck(false);
  }, [clearFilter]);
  return (
    <label className="checkbox">
      <span className="checkbox__input">
        <input
          type="checkbox"
          name="checkbox"
          checked={check}
          onChange={(event) => handleChange(event)}
        />
        <span className="checkbox__control">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 24 24"
            aria-hidden="true"
            focusable="false"
          >
            <path
              fill="none"
              stroke="white"
              strokeWidth="3"
              d="M1.73 12.91l6.37 6.37L22.79 4.59"
            />
          </svg>
        </span>
      </span>
      <span className="checkbox__label">{!label ? "none" : label}</span>
    </label>
  );
}

export default Checkbox;
