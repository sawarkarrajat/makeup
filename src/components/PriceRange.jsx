import React, { useState, useEffect } from "react";
import { useStateValue } from "./StateProvider";
import "../sass/PriceRange.sass";

/**
 * price range component
 * @property {function}
 */
function PriceRange() {
  const [state, dispatch] = useStateValue();
  const [minPrice, setMinPrice] = useState(state.priceMin);
  const [maxPrice, setMaxPrice] = useState(state.priceMax);
  /**
   * handles minimum price change
   * @param {Object} e
   */
  const handleSetMinPrice = (e) => {
    e.preventDefault();
    setMinPrice(e.target.value);
  };

  /**
   * handles maximum price change
   *
   * @param {Object} e
   */
  const handleSetMaxPrice = async (e) => {
    e.preventDefault();
    setMaxPrice(e.target.value);
  };

  /**
   * side effect to fire action to add price change to global state
   */
  useEffect(() => {
    let item = {
      min: minPrice,
      max: maxPrice,
    };
    dispatch({
      type: "ADD_PRICE_RANGE",
      item: item,
    });
  }, [minPrice, maxPrice, dispatch]);

  /**
   * side effect to reset prices
   */
  useEffect(() => {
    setMinPrice("");
    setMaxPrice("");
  }, [state.clearFilter]);

  /**
   * side effect to check if prices already exist in oldState of application
   */
  useEffect(() => {
    if (state.priceMin) {
      setMinPrice(state.priceMin);
    }
    if (state.priceMax) {
      setMaxPrice(state.priceMax);
    }
  }, [state.priceMin, state.priceMax]);
  return (
    <div className="priceRange__container">
      <div className="MuiTypography-root MuiTreeItem-label MuiTypography-body1">
        &#8226; price range
      </div>
      <div>
        <p className="MuiTypography-root MuiTreeItem-label MuiTypography-body1 priceRange__Subheading">
          &#8226; minimum
        </p>
        <input
          type="text"
          value={minPrice}
          onChange={(e) => handleSetMinPrice(e)}
          placeholder="&#36; 0.00"
          className="dashboard__searchInput priceRange__input"
        />
      </div>
      <div>
        <p className="MuiTypography-root MuiTreeItem-label MuiTypography-body1 priceRange__Subheading">
          &#8226; maximum
        </p>
        <input
          type="text"
          value={maxPrice}
          onChange={(e) => handleSetMaxPrice(e)}
          placeholder="&#36; 999.99"
          className="dashboard__searchInput priceRange__input"
        />
      </div>
    </div>
  );
}

export default PriceRange;
