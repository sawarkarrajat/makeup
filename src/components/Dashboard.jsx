import React, { useState, useRef } from "react";
import { toast } from "react-toastify";
import "../sass/Dashboard.sass";
import Navbar from "./Navbar";
import searchIcon from "../asset/search.png";
import crossIcon from "../asset/cross.png";
import check from "../asset/check.png";
import muData from "../makeupData.json";
import { useStateValue } from "./StateProvider";
import ClickAwayListener from "@material-ui/core/ClickAwayListener";
import ProductCard from "./ProductCard";
import LoadingSkeleton from "./LoadingSkeleton";
import girlImg from "../asset/girl.png";
import girlhhm from "../asset/girlhhmm.png";
import FilterTree from "./FilterTree";
import PriceRange from "./PriceRange";
import StarRating from "./StarRating";
import { ratingFilter } from "./filterServices";

// let muData = Object.assign([{}], makeupData);
const extractionLabels = (label) => {
  return [...new Set(muData.map((item) => item[label]))];
};

const extractionTaglist = () => {
  let tags = [];
  muData.forEach((item) => {
    item.tag_list.forEach((tag) => {
      if (!tags.includes(tag)) {
        tags.push(tag);
      }
    });
  });
  return tags;
};

const brand = extractionLabels("brand");
const tag_list = extractionTaglist();

const TypeSomething = () => {
  return (
    <div className="dashboard__msgContainer">
      <img src={girlImg} alt="girlmakeup" className="dashboard__girlImage" />
      <h3>
        type something in search bar above <br />
        to find products you might like...
      </h3>
    </div>
  );
};

const NothingFound = () => {
  return (
    <div className="dashboard__msgContainer">
      <h2>hhmm couldn't find anything...</h2>
      <img src={girlhhm} alt="girlmakeup" className="dashboard__girlImage" />
    </div>
  );
};

const autocompleteFilter = (searchedText) => {
  let matches = muData.filter((product) => {
    const regex = new RegExp(`^${searchedText}`, "gi");

    return (
      product.brand?.match(regex) ||
      product.name?.match(regex) ||
      product.category?.match(regex) ||
      product.product_type?.match(regex)
    );
  });
  return matches;
};
// const entityFilter = (products, entityName, value) => {
//   let data = products.filter((item) => item[entityName] === value);
//   console.log("value in entity filter", data);
// };

const Dashboard = () => {
  const [searchText, setSearchText] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [noResultsMsg, setNoResultsMsg] = useState(false);
  const [currentlyDisplayedCards, setCurrentlyDisplayedCards] = useState([]);
  const [suggestions, setSuggestions] = useState([]);
  const [
    { brandFiltersArray, tagFiltersArray, priceMin, priceMax, rating },
    dispatch,
  ] = useStateValue();
  const searchInput = useRef(null);
  const fakeLoading = () => {
    setIsLoading(true);
    setInterval(() => {
      setIsLoading(false);
    }, 2000);
  };

  const autoCompleteSearch = (e) => {
    e.preventDefault();
    let stext = e.target.value.trim(),
      matches;
    setSearchText(stext);
    if (stext.length === 0) {
      matches = [];
      setSuggestions([]);
      setNoResultsMsg(false);
      return;
    } else {
      matches = autocompleteFilter(stext);
      setSuggestions(matches);
    }
  };
  const searchSequence = () => {
    fakeLoading();
    if (suggestions.length === 0) {
      setNoResultsMsg(true);
    }
    setCurrentlyDisplayedCards(suggestions);
    setSuggestions([]);
  };
  const handleSearchClear = (e) => {
    e.preventDefault();
    setSearchText("");
    setSuggestions([]);
  };
  const handleSuggestionClick = (e, product) => {
    e.preventDefault();
    setSearchText(product.name);
    fakeLoading();
    setSuggestions([]);
    setCurrentlyDisplayedCards([product]);
  };
  const handleKey = (event) => {
    event.preventDefault();

    if (
      event.keyCode === 13 ||
      event.which === 13 ||
      event.key === "Enter" ||
      event.button === 0
    ) {
      searchSequence();
    }
  };
  const handleSearchButton = (e) => {
    e.preventDefault();
    searchSequence();
  };
  const handleClearFilters = (e) => {
    e.preventDefault();
    dispatch({
      type: "CLEAR_FILTER",
    });
  };
  const handleApplyFilters = async (e) => {
    e.preventDefault();
    if (currentlyDisplayedCards.length === 0) {
      toast("Search for a product to apply filters");
      searchInput.current.focus();
    } else {
      let products = Object.assign([{}], currentlyDisplayedCards);
      if (rating) {
        products = await ratingFilter(products, rating);
      }
      setCurrentlyDisplayedCards(products);
    }
  };
  return (
    <div className="dashboard__main">
      <Navbar />
      <div className="dashboard__container">
        <div className="dashboard__filter">
          <h3>filters</h3>
          <hr />
          <br />
          <div className="dashboard__filterLabel">
            <FilterTree treeLabel="brands" checkboxArray={brand} />
            <FilterTree treeLabel="tags" checkboxArray={tag_list} />
            <PriceRange />
            <StarRating />
          </div>
          <div className="dashboard__filterActions">
            <button
              className="dashboard__clearFilter dashboard__Button"
              onClick={(e) => handleClearFilters(e)}
            >
              clear all
              <img src={crossIcon} alt="search" />
            </button>
            <button
              className="dashboard__applyFilter dashboard__Button"
              onClick={(e) => handleApplyFilters(e)}
            >
              apply filter
              <img src={check} alt="filters" />
            </button>
          </div>
        </div>
        <div className="dashboard__aside">
          <div className="dashboard__searchPanel">
            <input
              type="text"
              value={searchText}
              ref={searchInput}
              onChange={(e) => autoCompleteSearch(e)}
              onKeyUp={(e) => {
                return handleKey(e);
              }}
              autoComplete="on"
              placeholder="search for products here..."
              className="dashboard__searchInput"
            />
            <button
              className="dashboard__searchButton dashboard__Button"
              onClick={(e) => handleSearchButton(e)}
            >
              search
              <img src={searchIcon} alt="search" />
            </button>
            <button
              className="dashboard__clearButton dashboard__Button"
              onClick={(e) => handleSearchClear(e)}
            >
              clear
              <img src={crossIcon} alt="cross" />
            </button>
          </div>
          {suggestions.length > 0 && (
            <ClickAwayListener onClickAway={() => setSuggestions([])}>
              <div className="dashboard__searchSuggestions">
                {suggestions.map((product) => (
                  <div
                    key={product.id}
                    className="dashboard__suggestion"
                    onClick={(e) => handleSuggestionClick(e, product)}
                  >
                    <p className="dashboard__suggestionName">{product.name}</p>
                    <p className="dashboard__suggestionCategory">
                      {!product.category ? "none" : product.category}
                    </p>
                    <p className="dashboard__suggestionBrand">
                      {!product.brand ? "no brand" : product.brand}
                    </p>
                  </div>
                ))}
              </div>
            </ClickAwayListener>
          )}
          {isLoading ? (
            <div className="dashboard__cardsContainer">
              <LoadingSkeleton />
            </div>
          ) : (
            <div className="dashboard__cardsContainer">
              {currentlyDisplayedCards?.length > 0 ? (
                currentlyDisplayedCards.map((product) => (
                  <ProductCard key={product.id} product={product} />
                ))
              ) : noResultsMsg ? (
                <NothingFound />
              ) : (
                <TypeSomething />
              )}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default Dashboard;
