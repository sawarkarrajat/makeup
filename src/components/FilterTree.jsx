import React from "react";
import "../sass/Dashboard.sass";
import TreeView from "@material-ui/lab/TreeView";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import ChevronRightIcon from "@material-ui/icons/ChevronRight";
import TreeItem from "@material-ui/lab/TreeItem";
import Checkbox from "./Checkbox";

/**
 * filter tree component to render filters
 *
 * @param {propType} props
 */
function FilterTree({ treeLabel, checkboxArray }) {
  return (
    <>
      <TreeView
        defaultCollapseIcon={<ExpandMoreIcon />}
        defaultExpandIcon={<ChevronRightIcon />}
      >
        <TreeItem nodeId={treeLabel} label={treeLabel}>
          {checkboxArray.map((filter) => (
            <Checkbox key={filter} filterLabel={treeLabel} label={filter} />
          ))}
        </TreeItem>
      </TreeView>
      <br />
    </>
  );
}

export default FilterTree;
