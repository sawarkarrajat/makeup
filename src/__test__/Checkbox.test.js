import React from "react";
import renderer from "react-test-renderer";
import Checkbox from "./../components/Checkbox";
import { StateProvider } from "../components/StateProvider";
import reducer, { initialState } from "../components/reducer";
import { fireEvent, render, screen } from "@testing-library/react";
// import { useStateValue } from "./../components/StateProvider";
const treeLabel = "brand",
  filter = "aqua";
// const [state, dispatch] = useStateValue();
it("renders correctly", () => {
  const tree = renderer
    .create(
      <StateProvider initialState={initialState} reducer={reducer}>
        <Checkbox filterLabel={treeLabel} label={filter} />
      </StateProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
it("checkbox label exist", () => {
  const { getByText, getByRole } = render(
    <StateProvider initialState={initialState} reducer={reducer}>
      <Checkbox filterLabel={treeLabel} label={filter} />
    </StateProvider>
  );

  expect(screen.getByText("aqua")).toHaveTextContent(filter);

  // fireEvent.click(getByText("aqua"));
  // const expected = ["aqua"];
  expect(screen.getByRole("checkbox")).not.toHaveAttribute("disabled");
});
