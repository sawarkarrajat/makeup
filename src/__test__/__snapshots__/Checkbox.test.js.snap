// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<label
  className="checkbox"
>
  <span
    className="checkbox__input"
  >
    <input
      checked={false}
      name="checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="checkbox__control"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.73 12.91l6.37 6.37L22.79 4.59"
          fill="none"
          stroke="white"
          strokeWidth="3"
        />
      </svg>
    </span>
  </span>
  <span
    className="checkbox__label"
  >
    aqua
  </span>
</label>
`;
